### ISSUE TITLE
[tool] flutter --version is miscalculated on master

### ISSUE THREAD

**Issue by CaseyHillers** [flutter/flutter#110536]
I see:

```
Flutter 3.1.0-0.0.pre.2565 • channel master • https://github.com/flutter/flutter.git
Framework • revision b4c0a125b6 (10 minutes ago) • 2022-08-29 14:13:08 -0700
Engine • revision ed6bb2e2e7
Tools • Dart 2.19.0 (build 2.19.0-147.0.dev) • DevTools 2.16.0
```

I would expect some variant of 3.3.0.pre.#### as that's what the latest beta is.

This is likely related to the dev deprecation.

**Comment by tvolkert** [comment]
FYI @christopherfujino

**Comment by zanderso** [comment]
At HEAD with a fresh checkout of master:
```
$ git describe --long --tags HEAD
3.0.0-0.0.pre-2565-gb4c0a125b6
```
I tried on both Linux and macOS in case there was a git version issue, but got the same result. It looks like something might have gone wrong with tagging since 3.0?

**Comment by christopherfujino** [comment]
This is because the branch point of the flutter-3.2-candidate.5 branch and master is not tagged: https://github.com/flutter/flutter/commits/flutter-3.2-candidate.5

The first tag was on a cherrypick: https://github.com/flutter/flutter/commit/f90da9b1672f7d006ba760fbb8fd1aa17af1a82a

**Comment by CaseyHillers** [comment]
Should we then make it policy that new betas don't contain CPs?

**Comment by CaseyHillers** [comment]
\cc @itsjustkevin

**Comment by christopherfujino** [comment]
> Should we then make it policy that new betas don't contain CPs?

The way this is supposed to work is that the conductor should detect when a new branch is being promoted to beta, and backtrack and tag the branch point: https://github.com/flutter/flutter/blob/master/dev/conductor/core/lib/src/start.dart#L424

However, it looks like that code doesn't run if a `versionOverride` is provided, I'm guessing that's what happened here? If a version override is provided, we can't reliably auto-back the version to tag the branch point (there may be a namespace collision at that version, or it might be a negative number).

**Comment by christopherfujino** [comment]
Maybe what we should do is, if the version override is provided, still check that the branch point is tagged, and if it is not throw a state error and the user will probably need to manually tag the branch point.

**Comment by itsjustkevin** [comment]
Thanks for the context @christopherfujino, assigning this to myself for the time being, will investigate solutions.

**Comment by itsjustkevin** [comment]
To verify, what we want here is: 
1. Test to make sure master > beta > stable.
2. Automatically increment some value on master when a new beta is release. 
As for point 2, which value of the master version do we want to be ahead?

**Comment by christopherfujino** [comment]
> Should we then make it policy that new betas don't contain CPs?

For context, this was how things worked with dev releases, unless something had been reverted as part of an internal roll.

**Comment by christopherfujino** [comment]
@itsjustkevin as an FYI, I just closed two duplicates of this https://github.com/flutter/flutter/issues/110652 & https://github.com/flutter/flutter/issues/110667. We could unblock users on master by adding a tag unassociated with a release to the latest candidate branch.

**Comment by christopherfujino** [comment]
Actually, if we had the internal roll tool always tag the branch point, that would simplify things, and then the conductor wouldn't have to backtrack.

**Comment by CaseyHillers** [comment]
I'm happy to update the internal roll to also tag since it creates the candidate branches, but what would it tag?

The candidate versions rarely align with the release versions.

**Comment by itsjustkevin** [comment]
Encountered https://github.com/flutter/flutter/issues/111529 on this weeks stable release.  This may still be an issue.

**Comment by CaseyHillers** [comment]
The linked issue is related to Conductor instead of the Flutter tool. Closing as flutter --version on master is working as intended.
